sdk/core/azure-core/inc/azure/core/etag.hpp

```diff
@@ -0,0 +1,194 @@
+       | W/"1"  | "1"    | no match          | match           |
+       | "1"    | "1"    | match             | match           |
+       +--------+--------+-------------------+-----------------+
+
+    // etag:                            //This is possible and means no etag is present
+    // etag:""
+    // etag:"*"                         //This means the etag is value '*'
+    // etag:"some value"                //This means the etag is value 'some value'
+    // etag:/W""                        //Weak eTag
+    // etag:*                           //This is special, means any etag
+    // If-Match header can do this
+    // If-Match:"value1","value2","value3"  // Do this if any of these match
+    //
+
+    */
+    bool ETagComparison(const ETag& left, const ETag& right) const
+    {
+
+      // If both values are empty then we consider the ETag equal
+      if (!left.m_value && !right.m_value)
+        return true;
+
+      if (!left.m_value || !right.m_value)
+        return false;
+
+      // Previous comparison checked that both are Weak or both are strong
+      //  Thuse we can decide which comparison to do based on one of them
+      if (!left.IsWeak())
+      {
+        // Strong ETag so comparison is that the values match character for character
+        return left.m_value.GetValue().compare(right.m_value.GetValue()) == 0;
+      }
+      else
+      {
+        // Left side is weak ETag
+        if (right.IsWeak())
+        {
+          return left.m_value.GetValue().compare(right.m_value.GetValue()) == 0;
+        }
+        // Trim the weak tag
+        auto leftLength = left.m_value.GetValue().length();
+
+        if (leftLength - 2 != right.m_value.GetValue().length())
+        {
+          return false;
+        }
+        return left.m_value.GetValue().compare(2, leftLength - 2, right.m_value.GetValue()) == 0;
+      }
+      return false;
+    }
+
+  protected:
+    /**
+     * @brief Construct an empty (null) #ETag.
+     */
+    ETag() {}
+
+  public:
+    /**
+     * @brief Construct a #ETag.
+     */
+    explicit ETag(std::string etag) : m_value(std::move(etag)) {}
+
+    /**
+     * @brief Copy assignment.
+     */
+    ETag& operator=(const ETag&) = default;
+
+    /**
+     * @brief Whether @Tag is present.
+     * @return `true` if @ETag has a value, `false` otherwise.
+     */
+    const bool HasValue() { return m_value.HasValue(); }
+
+    /*
+     * @brief Return the contained string value
+     * @return #std::string
+     */
+    const std::string& ToString()
+    {
+      if (!m_value.HasValue())
+      {
+        abort();
+      }
+      return m_value.GetValue();
+    }
+
+    /**
+     * @brief Compare with \p other @ETag for equality.
+     * @param other Other @ETag to compare with.
+     * @return `true` if @ETag instances are equal, `false` otherwise.
+     */
+    bool operator==(const ETag& other) const { return ETagComparison(*this, other); }
```

Comment:
This is the only callsite for `ETagComparison`; consider simply inlining the definition here to reduce the total number of function definitions.

Reply:
Not if you take my suggestion